lightning/impl/sag_fast.pyx

```diff
@@ -291,23 +312,61 @@ def _sag_fit(self,
             g[i] = -loss.get_update(y_pred, y[i])
             g_change = g[i] - g_old
 
-            # Update coefficient scale (l2 regularization).
-            w_scale[0] *= (1 - eta_alpha)
-            # Take care of possible underflows.
 
-            if w_scale[0] < 1e-9:
-                for j in xrange(n_features):
+            # line-search procedure, to be done every 5 iterations
+            # for details, see section 4.6 of Schmidt, M., Roux, N., & Bach, F. (2013).
+            # "Minimizing finite sums with the stochastic average gradient".
+            # arXiv Preprint arXiv:1309.2388
+            if adaptive_eta > 0 and t % line_search_freq == 0 and fabs(g[i]) > 1e-8:
+                if square_norm_X[i] == 0:
+                    # compute the norm if not done already
+                    for j in range(n_nz):
+                        square_norm_X[i] += (data[j] * data[j])
+                a = loss.loss(y_pred - g[i] * square_norm_X[i] / lipschitz, y[i])
+                b = loss.loss(y_pred, y[i]) - 0.5 * square_norm_X[i] * (g[i] * g[i]) / lipschitz
+                if a <= b :
+                    # condition is satisfied, decrease Lipschitz constant
+                    lipschitz /= line_search_scaling
+                else:
+                    # condition not satisfied, decrease step size
+                    lipschitz *= 2.0
+
```

Comment:
I was surprised not to see a loop that iterates until the condition is reached. Have you checked Mark's implementation? I guess since the condition is against a single training instance, it doesn't make sense to pay too much cost in the line search.